src/mem/cache/tags/partitioning_policies/way_pp.cc

```diff
@@ -0,0 +1,127 @@
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "mem/cache/tags/partitioning_policies/way_pp.hh"
+
+#include <algorithm>
+
+#include "base/logging.hh"
+#include "base/trace.hh"
+#include "params/WayPartitioningPolicy.hh"
+#include "way_allocation.hh"
+
+namespace gem5
+{
+
+namespace partitioning_policy
+{
+
+WayPartitioningPolicy::WayPartitioningPolicy
+(const WayPartitioningPolicyParams &params): BasePartitioningPolicy(params)
+{
+    // get cache associativity and check it is usable for this policy
+    const auto cache_assoc = params.cache_associativity;
+    assert(cache_assoc > 0);
+
+    // iterate over all provided allocations
+    for (const auto allocation: params.allocations) {
+        const auto alloc_id = allocation->getPartitionId();
+
+        // save way allocations in policy
+        for (const auto way: allocation->getWays()) {
+
+            // check if allocations are valid
+            if ((0 <= way) && (way < cache_assoc)) {
+                if (this->partitionIdWays[alloc_id].count(way) == 0) {
+                    this->partitionIdWays[alloc_id].emplace(way);
+                } else {
+                    // do not add duplicate allocation to policy and warn
+                    warn("Duplicate Way Partitioning Policy allocation for "
+                        "PartitionID: %d, Way: %d",
+                        alloc_id, way);
+                }
+            } else {
+                if (way >= cache_assoc) {
+                    fatal("Way Partitioning Policy allocation for "
+                        "PartitionID: %d, Way: %d cannot be fullfiled as "
+                        "cache associativity is %d",
+                        alloc_id, way, cache_assoc);
+                }
```

Comment:
I believe this could be written with less indentation by using fatal_if

Reply:
Switched to using `fatal_if` in [commit 55d8b4d](https://github.com/gem5/gem5/pull/765/commits/55d8b4d735005a76ea2d04ff4f142eba22976f73)